File: merkle-trie/src/main/java/net/consensys/cava/trie/RemoveVisitor.java
package net.consensys.cava.trie;

import net.consensys.cava.bytes.Bytes;
import net.consensys.cava.concurrent.AsyncResult;

final class RemoveVisitor<V> implements NodeVisitor<V> {
  private final AsyncResult<Node<V>> nullNodeResult = AsyncResult.completed(NullNode.instance());

  @Override
  public AsyncResult<Node<V>> visit(ExtensionNode<V> extensionNode, Bytes path) {
    return extensionNode.path().then(extensionPath -> {
      int commonPathLength = extensionPath.commonPrefixLength(path);
      assert commonPathLength < path.size() : "Visiting path doesn't end with a non-matching terminator";

      if (commonPathLength == extensionPath.size()) {
        return extensionNode.child().then(child -> child.accept(this, path.slice(commonPathLength))).then(
            extensionNode::replaceChild);
      }

      // The path diverges before the end of the extension, so it cannot match

      return AsyncResult.completed(extensionNode);
    });
  }

  @Override
  public AsyncResult<Node<V>> visit(BranchNode<V> branchNode, Bytes path) {
    assert path.size() > 0 : "Visiting path doesn't end with a non-matching terminator";

    byte childIndex = path.get(0);
    if (childIndex == CompactEncoding.LEAF_TERMINATOR) {
      return branchNode.removeValue();
    }

    return branchNode.child(childIndex).then(
        child -> child.accept(this, path.slice(1)).then(
            updatedChild -> branchNode.replaceChild(childIndex, updatedChild)));
  }

  @Override
  public AsyncResult<Node<V>> visit(LeafNode<V> leafNode, Bytes path) {
    return leafNode.path().then(leafPath -> {
      int commonPathLength = leafPath.commonPrefixLength(path);
      return (commonPathLength == leafPath.size()) ? nullNodeResult : AsyncResult.completed(leafNode);
    });
  }

  @Override
  public AsyncResult<Node<V>> visit(NullNode<V> nullNode, Bytes path) {
    return nullNodeResult;
  }
}

File: merkle-trie/src/main/kotlin/net/consensys/cava/trie/experimental/RemoveVisitor.kt
package net.consensys.cava.trie.experimental

import net.consensys.cava.bytes.Bytes
import net.consensys.cava.trie.CompactEncoding

internal class RemoveVisitor<V> : NodeVisitor<V> {

  override suspend fun visit(extensionNode: ExtensionNode<V>, path: Bytes): Node<V> {
    val extensionPath = extensionNode.path()
    val commonPathLength = extensionPath.commonPrefixLength(path)
    assert(commonPathLength < path.size()) { "Visiting path doesn't end with a non-matching terminator" }

    if (commonPathLength == extensionPath.size()) {
      val child = extensionNode.child()
      val updatedChild = child.accept(this, path.slice(commonPathLength))
      return extensionNode.replaceChild(updatedChild)
    }

    // The path diverges before the end of the extension, so it cannot match

    return extensionNode
  }

  override suspend fun visit(branchNode: BranchNode<V>, path: Bytes): Node<V> {
    assert(path.size() > 0) { "Visiting path doesn't end with a non-matching terminator" }

    val childIndex = path.get(0)
    if (childIndex == CompactEncoding.LEAF_TERMINATOR) {
      return branchNode.removeValue()
    }

    val updatedChild = branchNode.child(childIndex).accept(this, path.slice(1))
    return branchNode.replaceChild(childIndex, updatedChild)
  }

  override suspend fun visit(leafNode: LeafNode<V>, path: Bytes): Node<V> {
    val leafPath = leafNode.path()
    val commonPathLength = leafPath.commonPrefixLength(path)
    if (commonPathLength == leafPath.size()) {
      return NullNode.instance()
    }
    return leafNode
  }

  override suspend fun visit(nullNode: NullNode<V>, path: Bytes): Node<V> = NullNode.instance()
}
